File: src/Java/copying_from_JIRA.java
import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.client.util.store.FileDataStoreFactory;
import com.google.api.services.sheets.v4.Sheets;
import com.google.api.services.sheets.v4.SheetsScopes;
import com.google.api.services.sheets.v4.model.*;
import com.atlassian.jira.rest.client.api.JiraRestClient;
import com.atlassian.jira.rest.client.api.domain.Issue;
import com.atlassian.jira.rest.client.api.domain.SearchResult;
import com.atlassian.jira.rest.client.internal.async.AsynchronousJiraRestClientFactory;
import com.atlassian.util.concurrent.Promise;
import org.apache.commons.codec.binary.Base64;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.URISyntaxException;
import java.security.GeneralSecurityException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.concurrent.ExecutionException;

public class JiraToSheets {

    private static final String APPLICATION_NAME = "Your Application Name";
    private static final JsonFactory JSON_FACTORY = GsonFactory.getDefaultInstance();
    private static final String TOKENS_DIRECTORY_PATH = "tokens";

    /**
     * Global instance of the scopes required by this quickstart.
     * If modifying these scopes, delete your previously saved tokens/ folder.
     */
    private static final List<String> SCOPES = Collections.singletonList(SheetsScopes.SPREADSHEETS);
    private static final String CREDENTIALS_FILE_PATH = "/credentials.json";

    // Configuration
    private static final String JIRA_DOMAIN = "";
    private static final String USERNAME = "";
    private static final String API_TOKEN = "";
    private static final String GENERAL_FILTER_ID = ""; 
    private static final String TECH_DEBT_FILTER_ID = "";
    private static final List<String> LABELS = Arrays.asList("");
    private static final String SPREADSHEET_URL = "Google sheet URL";
    private static final String SHEET_NAME = "SheetName";

    public static void main(String... args) throws IOException, GeneralSecurityException, URISyntaxException, ExecutionException, InterruptedException {

        // 1. Build a new authorized API client service.
        final NetHttpTransport HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
        Sheets service = new Sheets.Builder(HTTP_TRANSPORT, JSON_FACTORY, getCredentials(HTTP_TRANSPORT))
                .setApplicationName(APPLICATION_NAME)
                .build();

        // 2. Retrieve the spreadsheet and sheet
        Spreadsheet spreadsheet = service.spreadsheets().get(spreadsheetIdFromUrl(SPREADSHEET_URL)).execute();
        Sheet originalSheet = getSheetByName(spreadsheet, SHEET_NAME);
        if (originalSheet == null) {
            throw new RuntimeException("Original sheet not found: " + SHEET_NAME);
        }

        // 3. Copy and rename the sheet
        CopySheetToAnotherSpreadsheetRequest copyRequest = new CopySheetToAnotherSpreadsheetRequest();
        copyRequest.setDestinationSpreadsheetId(spreadsheet.getSpreadsheetId());
        SheetProperties newSheetProperties = service.spreadsheets().sheets().copyTo(spreadsheet.getSpreadsheetId(), originalSheet.getProperties().getSheetId(), copyRequest).execute();
        String newSheetName = getFormattedDate(1); 
        updateSheetName(service, spreadsheet.getSpreadsheetId(), newSheetProperties.getSheetId(), newSheetName);

        // 4. Move the new sheet
        // (Implementation depends on the Google Sheets API - you might need to reorder sheets)

        // 5. Clear Data and Preserve Formulas
        // (Implementation depends on the Google Sheets API - clear data, then reapply formulas)

        // 6. Remove Strikethrough formatting
        // (Implementation depends on the Google Sheets API)

        // 7. Fetch and Write Jira Tickets
        fetchAndWriteTickets(service, spreadsheet.getSpreadsheetId(), newSheetProperties.getSheetId(), JIRA_DOMAIN, USERNAME, API_TOKEN, GENERAL_FILTER_ID, LABELS, 150);

        // 8. Fetch and Write Tech Debt Tickets
        fetchAndWriteTechDebtTickets(service, spreadsheet.getSpreadsheetId(), newSheetProperties.getSheetId(), JIRA_DOMAIN, USERNAME, API_TOKEN, TECH_DEBT_FILTER_ID);
    }

    // ... (Implement other functions like fetchAndWriteTickets, fetchAndWriteTechDebtTickets, callJiraApi,
    // getFormattedDate, etc. using Java libraries for Jira and Google Sheets API interaction)

}

File: src/Ruby/copying_from_JIRA.rb
require 'google/apis/sheets_v4'
require 'googleauth'
require 'googleauth/stores/file_token_store'
require 'jira-ruby'
require 'date'

# 1. Configuration (Store in a separate YAML or .env file and load)
jira_domain = ''
username = ''
api_token = ''
general_filter_id = '' 
tech_debt_filter_id = ''
labels = [''] 
spreadsheet_url = 'Google sheet URL'
sheet_name = 'SheetName'

# 2. Google Sheets Authentication (using googleauth gem)
#  - Set up OAuth2 credentials and authorization (refer to googleauth gem documentation)

service = Google::Apis::SheetsV4::SheetsService.new
service.client_options.application_name = 'YourAppName'
service.authorization = # ... your authorization object

# 3. Open Spreadsheet and Create New Sheet
spreadsheet = service.get_spreadsheet(spreadsheet_id_from_url(spreadsheet_url))
original_sheet = spreadsheet.sheets.find { |s| s.properties.title == sheet_name }
raise "Original sheet not found: #{sheet_name}" unless original_sheet

tomorrow = Date.today + 1
new_sheet_name = tomorrow.strftime('%b-%d-%Y')

# (You'll likely need to use the Google Sheets API to duplicate the sheet and set its name)

# 4. Clear Data and Preserve Formulas
range = "#{sheet_name}!A1:Z" # Adjust range as needed
response = service.get_spreadsheet_values(spreadsheet.spreadsheet_id, range)
values = response.values
formulas = values.map { |row| [row[7]] } # Assuming formulas are in column H (index 7)

# Clear data but keep header row
requests = [
  {
    update_cells: {
      range: {
        sheet_id: new_sheet.properties.sheet_id,
        start_row_index: 1, # Start from row 2 (below header)
        end_row_index: new_sheet.properties.grid_properties.row_count,
        start_column_index: 0,
        end_column_index: new_sheet.properties.grid_properties.column_count
      },
      fields: 'userEnteredValue'
    }
  }
]

service.batch_update_spreadsheet(spreadsheet.spreadsheet_id, {
    requests: requests
})

# Re-apply formulas to column H
# (You'll likely need to use the Google Sheets API to update specific cells with formulas)

# 5. Fetch and Write Jira Tickets (using 'jira-ruby' gem)
def fetch_and_write_tickets(sheet, jira_domain, username, api_token, filter_id, labels, max_results)
  # ... (implementation using jira-ruby to fetch issues and process data)

  # Update Google Sheet with rows (using Google Sheets API)
end

# 6. Fetch and Write Tech Debt Tickets
def fetch_and_write_tech_debt_tickets(sheet, jira_domain, username, api_token, filter_id)
  # ... (implementation using jira-ruby to fetch tech debt issue)

  # Append row to Google Sheet (using Google Sheets API)
end

# 7. Call Jira API (using 'jira-ruby' gem)
def call_jira_api(jira_domain, username, api_token, jql, max_results)
  options = { :username => username, :password => api_token }
  client = JIRA::Client.new(options)

  # Search for issues using JQL
  issues = client.Issue.jql(jql, max_results: max_results)
  return { 'issues' => issues } 
end

# 8. Main Execution
if __FILE__ == $0
  fetch_and_write_tickets(new_sheet, jira_domain, username, api_token, general_filter_id, labels, 150)
  fetch_and_write_tech_debt_tickets(new_sheet, jira_domain, username, api_token, tech_debt_filter_id)
end